### ISSUE TITLE
Doesn't work with with non sh-like shells

### ISSUE THREAD

**Issue by paride** [MisterTea/EternalTerminal#106]
When using a shell that doesn't allow the `VAR=val command` syntax, `et` fails to start `etserver`. This is the error message:

```
fish: Unsupported use of '='. In fish, please use 'set SERVER_TMP_DIR $TMPDIR:-$TMP:-$TEMP:-/tmp'.
SERVER_TMP_DIR=${TMPDIR:-${TMP:-${TEMP:-/tmp}}};TMPFILE=$(mktemp $SERVER_TMP_DIR/et-server.XXXXXXXXXXXX);PASSKEY=xxx;ID=yyy;printf "%s/%s\n" "$ID" "$PASSKEY" > "${TMPFILE}";export TERM=st-256color;etserver --idpasskeyfile="${TMPFILE}";true
```

and the fix probably consist in running `/bin/sh -c ...` instead of relying on the default shell.

**Comment by tbodt** [comment]
The solution is likely to change $SHELL to /bin/sh in two places in SshSetupHandler.cpp.

**Comment by ailzhang** [comment]
@MisterTea Do we want to replace current $SHELL with /bin/sh as suggested by @tbodt ? It makes sense to assume everyone has /bin/sh imho.

**Comment by MisterTea** [comment]
If we change the shell, the the path will be incorrect for people who rely on .bashrc or .zshrc to find etserver, no?

Unfortunately "set" means something different for bash & zsh, so to address this, we need to have two versions of the ssh script: one for fish and one for everyone else.

**Comment by ailzhang** [comment]
Yes you are right about the .bashrc and .zshrc. In interactive mode of ssh, it does find the PATH in those files. 

Although a bit messy in code, do we want to add the support for non sh-like shells? Looks like this is the only place that shell matters.

**Comment by MisterTea** [comment]
Fortunately we have a good solution: we just need to use export

**Comment by MisterTea** [comment]
Doh, looks like there's another issue:  ${VARIABLE} is not supported in fish

**Comment by tbodt** [comment]
Couldn't all of this stuff be done in etserver instead of with a shell script?

**Comment by MisterTea** [comment]
To do that we would need to link in our own ssh library that works on all platforms.  It's do-able but not easy.

**Comment by tbodt** [comment]
Would you? The only thing I see happening in the shell script is creating a temp file and writing a fixed string into the temp file.

**Comment by MisterTea** [comment]
oh, I see what you are saying.  We could just call etserver and let etserver create the passkey and print it to stdio.  That's a great idea

**Comment by MisterTea** [comment]
Hmm, actually that won't work because we need the client to create the id & passkey on the client and pass it in securely.  If I make it a command line argument then anyone can read it from ps

**Comment by MisterTea** [comment]
It's hard (impossible?) to write some if/then/else that runs a fish script or a bash script.  I tried this:

```
 fish -lc 'printf "${0}"' || echo "failed"
```

but even though the fish command failed, it didn't return exit code 127 as advertised. Someone who knows more about fish than me will have to write some shell script that runs one set of commands for fish and a different set for everyone else.

Alternatively, you can run fish by doing  et -c="fish" (hostname).

I'm going to close this because I don't expect anyone to implement it, but if someone does and submits a PR I'd be happy to reopen.

**Comment by paride** [comment]
A user can have any kind of shell, I think it's wrong to assume anything more than the existance of `/bin/sh`. If you recognize this as an issue, and it is not fixed, I think it should be left open.

**Comment by MisterTea** [comment]
I'll keep it open for a few months but if no one picks it up I'll close it.

**Comment by tbodt** [comment]
You could send it through stdin with echo and a pipe, both of those work in fish.

**Comment by vvug** [comment]
This project can't be considered serious if it silently depends on a user's shell choice. @tbodt solution makes sense. If `etserver` is not in the PATH, then allow specifying its full path as an etclient option.

**Comment by tbodt** [comment]
@qolii I have my shell set to bash and this in my bashrc:

```bash
if [ -z "$BASH_EXECUTION_STRING" ]; then
    exec /usr/bin/fish --login
fi
```

**Comment by phiber1** [comment]
% et alien
phiber@alien's password:
Unknown option: `-lc'
Usage: tcsh [ -bcdefilmnqstvVxX ] [ argument ... ].
Error starting ET process through ssh, please make sure your ssh works first

Not everyone uses the same shell, as illustrated several times.

**Comment by paride** [comment]
For reference: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=865881#10.

**Comment by MisterTea** [comment]
Here is the code that runs server-side to establish the connection:

https://github.com/MisterTea/EternalTerminal/blob/master/src/terminal/SshSetupHandler.cpp#L26

If someone on this thread can change that logic to support fish, I would be more than happy to review a PR.

**Comment by MisterTea** [comment]
Actually I may have a solution here: we do ssh -> (user shell) -> sh -> script.  This way the PATH and other environment is set correctly, but we do not depend on the user shell's interpreter.

**Comment by MisterTea** [comment]
For example:

```ssh localhost "$SHELL -lic \" sh -c \\\" echo \\\$PATH \\\" \" "```

**Comment by paride** [comment]
I don't think you can assume `-lic` are valid options for `$SHELL`. Personally I'd do whatever mosh is doing with its `--server=COMMAND` client side option.

**Comment by phiber1** [comment]
FWIW, the -l, -i, and -c options are standard and have the exact same
meaning across sh, csh, tcsh, ksh, bash, and zsh. Probably others.
But I agree, the server side should not be dependent on $SHELL.

-Mark

**Comment by tbodt** [comment]
-l, -i, -c also work for fish.

**Comment by MisterTea** [comment]
@paride That may work, but you would need to move all of the shell script logic into the binary first before you could support that.

**Comment by paride** [comment]
The `-lic` options are pretty widespread but not ubiquitous, for example they are not supported by cicada (https://github.com/mitnk/cicada) and probably by ion (https://github.com/redox-os/ion, but I didn't check). I still think there should be no dependency on `$SHELL`.

What about falling back to a standard location if `etserver` is not found in `$PATH`, e.g. `~/bin/etserver`?

**Comment by paride** [comment]
Or call `sh -lc`, which will read the `~/.profile` file, which is also read by bash and zsh.

**Comment by ailzhang** [comment]
@paride When I tried `sh -lc` on mac, it doesn't trigger `/etc/profile`, while it does correctly on Linux. Do you know why it happens?

**Comment by ailzhang** [comment]
@paride Yea that's basically the main blocker from using `sh` directly.
We will likely go with @MisterTea 's solution for now and come back when people really see problems with `lic`.

**Comment by phiber1** [comment]
Bear in mind that on csh/tcsh, there is no .profile file. All shells read
.cshrc, login shells in addition read .login, and the -l option forces
.login to be read.
Also, tcsh can use .tcshrc instead of .cshrc, so that csh and tcsh can
co-exist, as advanced tcsh commands in a .cshrc would otherwise give csh
errors.
But I digress.

-Mark

**Comment by paride** [comment]
I am not sure I understood why `export PATH` can't be used. 

What about using `/usr/bin/env`? Something like: `env PATH=$PATH sh -c "..."`.

**Comment by ailzhang** [comment]
I think `fish` doesn't support `export` iirc. Instead they use `set`.  @paride

**Comment by paride** [comment]
@ailzhang `fish` does support `export` (but it's not only about `fish` here!).

**Comment by phiber1** [comment]
Also no such thing as "export" on csh/tcsh.
The equivalent is "setenv" for the global environment, and you don't use an
equal sign.
"set" is for local environment, and you *do* use an equals.

**Comment by paride** [comment]
Then what remains is `env PATH=$PATH /bin/sh -c "..."` (were we assume: a `PATH` variable accessible as `$PATH`).

**Comment by phiber1** [comment]
Confirmed that works fine on csh/tcsh.

**Comment by paride** [comment]
I retreat: `$PATH` in `fish` has a space as delimiter, so it is not compatible with `sh` :(

**Comment by paride** [comment]
I guess this leaves us with `$SHELL -lic`, maybe `$SHELL -lc` could be enough and a bit more general. I would still prefer a hardcoded fallback path if `etserver` is not found, but this is just my opinion.

**Comment by MisterTea** [comment]
https://github.com/MisterTea/EternalTerminal/tree/add_fish_support  

This branch adds support for fish and should also support almost any shell.  It uses the idea mentioned on this thread to pass in the data through stdin.

Because this changes the protocol between client/server, both of them need to be running this branch for it to work.

Can someone with fish confirm that this works?

**Comment by paride** [comment]
@MisterTea I did test it very quickly and it seems to be working fine!

**Comment by MisterTea** [comment]
Cool, I'll make a PR and it will go out in our next release

**Comment by MisterTea** [comment]
https://github.com/MisterTea/EternalTerminal/pull/140

**Comment by qolii** [comment]
Also, FYI, this works great for me! Thanks for doing this!

**Comment by phiber1** [comment]
This doesn't work with tcsh. For starters, the -l option can only be used
alone, with no other options, unfortunately.
The add_fish_support branch returns the following from etserver:
Unknown option: `-lc'
Usage: tcsh [ -bcdefilmnqstvVxX ] [ argument ... ].

-Mark

**Comment by MisterTea** [comment]
@phiber1 Are you sure you are using the right branch?  The -lc option is gone in the fish branch.

**Comment by phiber1** [comment]
Initially I had only updated the server side, and got that error on the
(original) client.

Just now I updated the client to also be the fish branch, but I get a
different error, right after typing my password:

Unmatched '"'.
Unmatched '"'.
Error starting ET process through ssh, please make sure your ssh works first

**Comment by qolii** [comment]
I... appear to be seeing this now, with the 5.1.6 release (on both server and client).
```
$ et qoli@xxx.org
Error starting ET process through ssh, please make sure your ssh works first

$ ssh qoli@xxx.org
Last login: Wed Oct 17 09:30:50 2018 from 11.22.33.44
Welcome to fish, the friendly interactive shell
Type help for instructions on how to use fish
[qoli@braid ~$]
```
Is there a specific kind of log or data I can collect that might help?

**Comment by MisterTea** [comment]
The fish changes are in the 6.0 branch so fish still won't work on 5.1.6, which is a patch release for the 5.0 branch.

**Comment by qolii** [comment]
Ohh! Silly me :) Hopefully 6.0 can come soon!

**Comment by qolii** [comment]
(And thanks for the quick update!)

**Comment by wilchak** [comment]
I am still experiencing this bug, even with a freshly compiled "add_fish_support" branch. When running et on the client, I get this:
`et myuser@myserver.com -logtostdout
fish: Expected end of the statement, but instead found a string
SERVER_TMP_DIR=${TMPDIR:-${TMP:-${TEMP:-/tmp}}};TMPFILE=$(mktemp $SERVER_TMP_DIR/et-server.XXXXXXXXXXXX);PASSKEY=YaS6o28jo1Mr2qiZsQ4uUeNo5Y0IfWO0;ID=0ej8lbInJA9h51Ma;printf "%s/%s\n" "$ID" "$PASSKEY" > "${TMPFILE}";export TERM=xterm-256color;if [ -x "$(command -v etterminal)" ]; then  etterminal --idpasskeyfile="${TMPFILE}" --v=0; else  etserver --idpasskeyfile="${TMPFILE}" --v=0;fi;true
                                                                                                                                                                                                                                                                                                                                                   ^

Error starting ET process through ssh, please make sure your ssh works first`

**Comment by mdemoney** [comment]
I'm a tcsh user and would like to use ET, but am blocked by this bug.

I tried this experiment: /usr/bin/ssh somehost '/bin/sh -lc "FOO=bar /usr/bin/printenv"'

Interestingly, this works and prints my full tcsh path plus the added FOO value.  I do not have a .profile or .bash* on the remote machine.  I don't have any SendEnv commands in my ssh_config.

I realize that the ssh manpage says that command is run _in place_ of the login shell, but by this experiment, it seems to be run in the user's chosen shell.

Could replacing $SHELL with /bin/sh in SshSetupHandler.cpp be a quick route to support for other shells until a longer term fix is available?

Even, if I'm misinterpreting something here, at least this change would allow those of us on other shells to use ET by just adding a .profile with the appropriate path to find etserver.

**Comment by MisterTea** [comment]
@mdemoney The v6 branch should work with tsch?

**Comment by MisterTea** [comment]
@mdemoney The v6 branch should work with tsch?

**Comment by mdemoney** [comment]
Thanks, I'll definitely build that and give it a try on the hosts where I have root.
(A released fix is an easier sell to admins of hosts that I don't control. Hopefully, v6 will release soon.)

**Comment by mdemoney** [comment]
Tried building v6, ran into this problem:
Scanning dependencies of target et-test
[ 92%] Building CXX object CMakeFiles/et-test.dir/test/BackedTest.cpp.o
In file included from /dbc/pa-dbc1112/mdemoney/src/et/EternalTerminal/test/BackedTest.cpp:1:0:
/dbc/pa-dbc1112/mdemoney/src/et/EternalTerminal/test/TestHeaders.hpp:5:28: fatal error: catch2/catch.hpp: No such file or directory
 #include "catch2/catch.hpp"

The v6 cmake didn't seem to have a BUILD_TEST option to disable.

FWIW: I did get the 5.1.9 working with tcsh.  Changed both $SHELL in SshSetupHandler.cpp to /bin/sh and also found a couple places passing --login to the shell (TerminalHandler.cpp and UserTerminalHandler.cpp) which I replaced with -l.

**Comment by MisterTea** [comment]
Catch is a new git submodule in v6 so you need to pull it. 

If you have 5.1.9 working, send us a PR!  The catch is that it can't break bash/zsh which are the two most popular shells.

**Comment by MisterTea** [comment]
Actually I re-read your comment and that approach won't work for people who install et to a non-standard directory and needs to load their Shell's profile to get that path (this effects all OS/X servers).

Please try the v6 branch and make sure it works, thanks!

**Comment by mdemoney** [comment]
Hi, thanks for the response.  I'll definitely give the v6 branch a try with tcsh.

FWIW: I realized that "ssh HOST '/bin/sh -lc ....' works for me (even on OS/X servers) because my shell rc files initialize the environment (if not already setup) even for non-login shells (precisely because I wanted "ssh HOST command" to work for commands in ~/bin).  Of course, that's not a solution for everyone.

**Comment by chriseomi** [comment]
```bash
chrismac:~% $(echo $0) --version
GNU bash, version 4.4.12(1)-release (x86_64-apple-darwin16.3.0)
Copyright (C) 2016 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>

This is free software; you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
chrismac:~% et --version
et version 5.1.9
chrismac:~% et
Error starting ET process through ssh, please make sure your ssh works first
```

Just installed ET. Not working for me and getting error `Error starting ET process through ssh, please make sure your ssh works first`. No idea about any of this. Using iTerm on latest macos.

My ssh works just fine, but is aliased in bash.  `ssh is aliased to `ssh -F ~/.ssh/config -i ~/.ssh/id_ed25519 -o UserKnownHostsFile=~/.ssh/known_hosts '`

Not sure if that affects it anything.

**Comment by MisterTea** [comment]
@chriseomi Is the et server running?

**Comment by chriseomi** [comment]
@MisterTea yup. Also I removed known_hosts, and see it gave me the usual SSH warning about fingerprints.  IOW, it's definitely invoking SSH.

Couldn't see an obvious way to do verbose logging to help diagnose

**Comment by MisterTea** [comment]
v6 adds support for fish, tcsh, and most other shells.

**Comment by benjaminmordaunt** [comment]
@MisterTea 
Um. I'm not sure this is true. I'm running 6.1.8 on both client and server and am still seeing:
```
Unmatched '''.
Unmatched '''.
Error starting ET process through ssh, please make sure your ssh works first
```
right after I enter my password...

**Comment by fpaquet** [comment]
Same problem here, with tcsh
RHEL8, et 6.1.9
CentOS7, et 6.0.11

When switching login shell to /bin/sh, both are working

**Comment by brechtm** [comment]
Running into the same issue with tcsh. Unfortunately, I'm in an environment where I cant change my default shell.

Changing the "default SSH shell" adding a _RemoteCommand_ line in `.ssh/config` doesn't seem to be supported by ET as it print's the following:
```
Cannot execute command-line and remote command.
```

**Comment by brw** [comment]
I'm also in an environment where I can't change my default shell, so I thought I'd change my by doing `exec fish -l` in my .bashrc. However now when I attempt to connect to my machine using `et` it just hangs forever until I `Ctrl-C`.
